make_fx: include fqns in the graph when running make_fx on nn modules

bdhirsh opened the pull request:
In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103026
* __->__ #103025
* #103024

bdhirsh committed:
make_fx: include fqns in the graph when running make_fx on nn modules

[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

ezyang commented:
I'd defer to @Chillee on this one, he has opinions about how much we should work to preserve fqns on make_fx

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

Chillee commented:
Hmm... I think this is a good idea.

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]

bdhirsh committed:
Update on "make_fx: include fqns in the graph when running make_fx on nn modules"

In retrospect, this PR isn't really necessary - left it up because I added the functionality, and it seems marginally useful - but I'm happy to can it for now if we don't want to land it.

The idea was that I wanted to be able to access parameters from a module, after performing pre_dispatch tracing on it. e.g.:
```
mod = MyModule()
graph = make_fx(mod, pre_dispatch=True)(*args)
# Do something useful
print([(k, v.shape) for k, v in graph.named_parameters()])
```

It turns out that this already "works", partially. When `torch.fx.tracer.trace()` sees a piece of non-local state during tracing, it will automatically assign it as a parameter on the output `GraphModule`, if it sees that the state is an `nn.Parameter` ([here](https://github.com/pytorch/pytorch/blob/4479e2fa193c06a240575668e5753f3a3c741917/torch/fx/graph_module.py#L209)). We lose fqn's though: it just assigns the parameter name as `_tensor_constantN`.

This PR basically just attempts to preserve fqn's, by mangling periods into underscores. In theory, someone who cared about fqn's, and wanted to do different things to different parameters depending on their fqn, could demangle them (e.g. if they wanted to do the same thing to every parameter in a submodule). Example (also see the expecttest in the PR):
```
        self.assertExpectedInline(fx_g.print_readable(print_output=False), """\
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 3, 3, 3]):
        # No stacktrace found for following nodes
        bn_weight = self.bn_weight
        bn_bias = self.bn_bias
        bn_running_mean = self.bn_running_mean
        bn_running_var = self.bn_running_var
        bn_num_batches_tracked = self.bn_num_batches_tracked
        add_: i64[] = torch.ops.aten.add_.Tensor(bn_num_batches_tracked, 1);  bn_num_batches_tracked = None
        batch_norm: f32[3, 3, 3, 3] = torch.ops.aten.batch_norm.default(arg0_1, bn_weight, bn_bias, bn_running_mean, bn_running_var, True, 0.1, 1e-05, True);  arg0_1 = bn_weight = bn_bias = bn_running_mean = bn_running_var = None
        return batch_norm
        """)
```




[ghstack-poisoned]
